File: src/Graphics/Animation.cpp
// Copyright (c) 2010-present Bifrost Entertainment AS and Tommy Nguyen
// Distributed under the MIT License.
// (See accompanying file LICENSE or copy at http://opensource.org/licenses/MIT)

#include "Graphics/Animation.h"

#include "Director.h"

using rainbow::Animation;
using rainbow::SpriteRef;

Animation::Animation(const SpriteRef& sprite,
                     Frames frames,
                     unsigned int fps,
                     int delay)
    : interval_(1000 / fps), frames_(std::move(frames)), sprite_(sprite),
      delay_(delay)
{
}

void Animation::set_frame_rate(unsigned int fps)
{
    const float progress = narrow_cast<float>(accumulated_) / interval_;
    interval_ = 1000 / fps;
    accumulated_ = interval_ * progress;
}

void Animation::set_frames(Frames frames)
{
    frames_ = std::move(frames);
    rewind();
}

void Animation::set_sprite(const SpriteRef& sprite)
{
    sprite_ = sprite;
    if (!is_stopped())
        set_current_frame();
}

void Animation::jump_to(unsigned int frame)
{
    // TODO: Check for invalid frame.
    accumulated_ = 0;
    frame_ = frame;
    idled_ = 0;
    set_current_frame();
}

void Animation::start()
{
    if (!is_stopped())
        return;

    stopped_ = false;
    rewind();
    if (callback_)
        callback_(this, AnimationEvent::Start);
}

void Animation::stop()
{
    if (is_stopped())
        return;

    stopped_ = true;
    if (callback_)
        callback_(this, AnimationEvent::End);
}

void Animation::update(uint64_t dt)
{
    if (is_stopped())
        return;

    accumulated_ += std::min<uint32_t>(dt, interval_ * 5);
    while (accumulated_ >= interval_)
    {
        // Consume first as |accumulated_| may be reset during a tick, leading
        // to an underflow.
        accumulated_ -= interval_;
        tick();
    }
}

void Animation::set_current_frame()
{
    if (!frames_ || !sprite_)
        return;

    sprite_->texture(frames_[frame_]);
}

void Animation::tick()
{
    if (frames_[frame_ + 1] == end_frame())
    {
        if (delay_ < 0)
        {
            stop();
            return;
        }
        if (idled_ == 0 && callback_)
        {
            callback_(this, AnimationEvent::Complete);
            if (is_stopped())
                return;
        }
        if (idled_ < delay_)
        {
            ++idled_;
            return;
        }
        frame_ = 0;
        idled_ = 0;
    }
    else
    {
        ++frame_;
        if (callback_)
            callback_(this, AnimationEvent::Frame);
    }
    sprite_->texture(frames_[frame_]);
}

#ifndef NDEBUG
Animation::~Animation()
{
    Director::assert_unused(
        this, "Animation deleted but is still in the render queue.");
}
#endif

File: dotnet/Rainbow/Graphics/Animation.cs
// Copyright (c) 2010-present Bifrost Entertainment AS and Tommy Nguyen
// Distributed under the MIT License.
// (See accompanying file LICENSE or copy at http://opensource.org/licenses/MIT)

namespace Rainbow.Graphics
{
    using System;
    using System.Diagnostics.Contracts;
    using System.Runtime.InteropServices;

    using Interop;

    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    internal delegate void AnimationEventHandler(
        IntPtr nativeHandle,
        AnimationEvent ev,
        uint currentFrame,
        IntPtr context);

    public sealed class Animation : NativeObject
    {
        private Sprite sprite;
        private int delay;
        private GCHandle gcHandle;
        private AnimationEventHandler eventHandler;

        public Animation(Sprite sprite, uint[] frames, uint fps, int delay = 0)
        {
            Contract.Requires(sprite != null);
            Contract.Requires(frames != null);
            Contract.Requires(frames.Length > 0);
            Contract.Requires(fps > 0);
            Contract.Ensures(NativeHandle != IntPtr.Zero);
            Contract.Ensures(this.sprite != null);

            NativeHandle = Core.AnimationCreate(
                sprite.Parent.NativeHandle,
                sprite.Index,
                frames,
                frames.GetSize(),
                fps,
                delay);
            this.sprite = sprite;
            this.delay = delay;
            this.gcHandle = GCHandle.Alloc(this, GCHandleType.Weak);
        }

        ~Animation()
        {
            gcHandle.Free();
        }

        public uint CurrentFrame => Core.AnimationCurrentFrame(NativeHandle);

        public int Delay
        {
            get
            {
                return delay;
            }

            set
            {
                delay = value;
                Core.AnimationSetDelay(NativeHandle, value);
            }
        }

        public uint FrameRate
        {
            get { return Core.AnimationFrameRate(NativeHandle); }
            set { Core.AnimationSetFrameRate(NativeHandle, value); }
        }

        public bool IsStopped => Core.AnimationIsStopped(NativeHandle);

        public Sprite Sprite
        {
            get
            {
                return sprite;
            }

            set
            {
                Contract.Requires(value != null);

                if (value == sprite)
                {
                    return;
                }

                sprite = value;
                Core.AnimationSetSprite(NativeHandle, value.Parent.NativeHandle, value.Index);
            }
        }

        protected override Action<IntPtr> Delete => Core.AnimationRelease;

        public void JumpTo(uint frame) => Core.AnimationJumpTo(NativeHandle, frame);

        public void Rewind() => JumpTo(0);

        public void SetEventHandler(EventHandler<AnimationEventArgs> handler)
        {
            if (handler == null)
            {
                eventHandler = null;
            }
            else
            {
                GCHandle gcHandleCopy = gcHandle;
                eventHandler = (nativeHandle, ev, currentFrame, _) =>
                {
                    var animation = gcHandleCopy.Target as Animation;
                    if (animation == null)
                    {
                        return;
                    }

                    Contract.Assert(nativeHandle == animation.NativeHandle);

                    var eventArgs = new AnimationEventArgs(ev, currentFrame);
                    handler.Invoke(animation, eventArgs);
                };
            }

            Core.AnimationSetEventHandler(NativeHandle, eventHandler, IntPtr.Zero);
        }

        public void SetFrames(uint[] frames)
        {
            Contract.Requires(frames != null);
            Contract.Requires(frames.Length > 0);

            Core.AnimationSetFrames(NativeHandle, frames, frames.GetSize());
        }

        public void Start() => Core.AnimationStart(NativeHandle);

        public void Stop() => Core.AnimationStop(NativeHandle);

        protected override void OnDisposing()
        {
            SetEventHandler(null);
        }
    }
}
